amida-tech/recruitment-registry
# Search API endpoint

#### What's this PR do?
Create a new survey search endpoint. It takes an object of search criteria, and returns the number of users whose survey answers' match those criteria. The search criteria support all single/multi answer types. See the swagger docs for the full endpoint documentation, but an example is given below:

`POST http://localhost:9005/api/v1.0/surveys/19/search`

```json
{
  "questions": [
    {
      "id": 173,
      "answer": {
        "feetInchesValue": {
          "feet": 5,
          "inches": 0
        }
      }
    },
    {
      "id": 174,
      "answer": {
        "bloodPressureValue": {
          "systolic": 101,
          "diastolic": 71
        }
      }
    },
    {
      "id": 175,
      "answer": {
        "choices": [
          {
            "id": 252,
            "monthValue": "05"
          },
          {
            "id": 253,
            "dayValue": "15"
          }
        ]
      }
    }
  ]
}
```

```json
{
  "count": 1
}
```

#### Related JIRA tickets:
https://jira.amida-tech.com/browse/RR-52

#### How should this be manually tested?
Run the survey model and integration tests. Run a custom search query and manually verify the resultant count is correct from the database. The example query above can be run on the test database after the survey integration spec runs.

#### Any background context you want to provide?
Writing this to run efficiently required a raw SQL query with a dynamic number of subqueries. `queryrize` was extended to support this.

#### Screenshots (if appropriate):
